src/projects/trio/trio.cpp

```diff
@@ -0,0 +1,232 @@
+                haplotypes[e_label].haplotype = decision;
+                haplotypes[alt_label].haplotype = other_haplo(decision);
+                count ++;
+            }
+        }
+    }
+    logger_.info() << "Updated " << count << " ambiguous bulges";
+}
+
+void HaplotypeRemover::removeHaplotype() {
+    size_t removed = 0;
+    size_t bridges = 0;
+    size_t removed_len = 0;
+    bool changed = true;
+    while (changed) {
+        changed = false;
+        std::vector<int> eids;
+        for (auto p: mg.edges) {
+            eids.push_back(p.first);
+        }
+        for (auto eid:eids){
+            if (mg.edges.find(eid) == mg.edges.end())
+                continue;
+            auto label = mg.edges[eid]->getLabel();
+            if (haplotypes.find(label) != haplotypes.end()) {
+                if (haplotypes[label].haplotype == haplotype_) {
+                    if (mg.edges[eid]->isBridge() && mg.edges[eid]->size() < SAVED_BRIDGE_CUTOFF) {
+                        bridges ++;
+                        logger_.info() << "Skipping edge " << eid << " as bridge\n";
+                        continue;
+                    }
+                    removed_len += mg.edges[eid]->size();
+                    deleteEdgeHaplo(eid);
+
+                    logger_.trace() << "removing " << eid  << " label " << label << endl;
+                    removed ++;
+                    changed = true;
+                } else { 
+                    logger_.trace() << "skipping edge label" << label <<" "<< haplotypes[label].haplotype << endl;
+                }
+            } else {
+                logger_.trace() << "skipping edge NOT FOUND label" << label << endl;
+                
+            }
+        }
+    }
+    logger_.info() << "Saved " << bridges << "bridges\n";
+    logger_.info() << "Removed " << removed << " edges of haplo " << haplotype_  << " total len " << removed_len << endl;
+}
+
+void HaplotypeRemover::process() {
+    updateFixableHaplotypes();
+    updateAmbiguousHaplotypes();
+    removeHaplotype();
+    logger_.debug() << "removed \n";
+    mg.printEdgeGFA(out_dir / "before_clean.gfa", true);
+    cleanGraph();
+    mg.printEdgeGFA(out_dir / "mdbg.hpc.gfa", true);
+
+}
+
+
+std::experimental::filesystem::path simplifyHaplo(logging::Logger &logger, size_t threads,
+                                                  const std::experimental::filesystem::path &output_file,
+                                                  const std::experimental::filesystem::path &diplo_graph,
+                                                  const std::experimental::filesystem::path &haployak,
+                                                  const char haplotype,  const std::experimental::filesystem::path &corrected_reads,
+                                                  io::Library & reads,   const std::experimental::filesystem::path &dir) {
+    multigraph::MultiGraph mmg;
+    mmg.LoadGFA(diplo_graph, true);
```

Comment:
Why do we read the graph and haployak separatly for each haplome? Yes we need to separately process them but we should just make two copies after we read all the information. Yes, we can not copy multigraph but it is actually a good idea to write a copy method. You can also easily get another copy by running DBG method the second time.

Reply:
Discussed, added todo